joeynmt/training.py

```diff
@@ -227,6 +257,7 @@ def train_and_validate(self, train_data: Dataset, valid_data: Dataset) \
                                     train=True, shuffle=self.shuffle)
         for epoch_no in range(self.epochs):
             self.logger.info("EPOCH %d", epoch_no + 1)
+            self.last_epoch += 1
```

Comment:
Why is this variable needed in addition to `epoch_no` counter?

Reply:
self.last_epoch is removed, "epoch_no" is instead passed to _save_checkpoint() as part of the name of a saved checkpoint file to distinguish checkpoints by save_freq or by validation_freq